### DISCUSSION TITLE
Can't return MongoIds from +page.server.js load function

### DISCUSSION THREAD

**Discussion by NormandoHall** [sveltejs/kit#7470] [Q&A] [answered]
When I query my Db to get an array of users, all comes with it's MongID. 

I get the error:

`Data returned from `load` while rendering /persons is not serializable: Cannot stringify arbitrary non-POJOs (data.rows[0]._id)`

So, to workaround this issue I need to iterate all over the array and convert each `_id.toString()`.

I think this is not good. I am missing something?

Regards

**Comment by infovore** [comment, marked as answer]
As the error says, you can't return non-POJO objects from `+page.server.js`. That is to say: if it's an object that also has methods/prototype methods on it, you get this error. That's because it's a client/server interaction - the `.server*` page runs in a Node environment, not a browser environment, and sends data over the network, so of course all it can return is plain JSON. 

To quote the [docs](https://kit.svelte.dev/docs/load#shared-vs-server):

> A server-only load function must return data that can be serialized with [devalue](https://github.com/rich-harris/devalue) — anything that can be represented as JSON plus things like BigInt, Date, Map, Set and RegExp, or repeated/cyclical references — so that it can be transported over the network.

Question: why do you think this is _not good_ - is it because you require the functionality of the Id on the front-end? What would you expect to happen? And how have you coped before sending Mongo data from server to client, and could that inform this?

One straightforward way to turn a richer object into a POJO - preserving its data, but stripping its functionality - is `JSON.parse(JSON.stringify(foo))`. It's never pretty, but it does work. However, I don't see how something like:

```
return myData.map(d => ({
  ...d,
  _id: d._id.toString()
})
```

(or similar) from your `load` function is actually that bad. You say "iterate over the array" like it's a bad thing, but functional transformations with stuff like `map` feel like a fairly reasonable thing to do?

**Reply by NormandoHall** [reply]
Thanks for the reply.

Before SK updates, I could directly return the contents of the array without needing to iterate over it. Not only do I have the `_id `field, but also other fields contain an ObjectId from mongoDB.

That's exactly what I do with map, but I find it annoying to do it on every query.

I don't need the functionality of the `_id` object, just its value in plain text.

I don't understand why devalue doesn't convert anything that isn't a POJO to a string.

Thanks anyway.

**Reply by infovore** [reply]
Well - before it was silently stripping functionality and just ignoring all the bits it couldn't return. Now it's telling you it can't, actually, send the ting you have over the wire, because there would be loss. In this case, you don't care what's lost, but that's not always true, and it's asking you to more explicitly send data that can actually be sent. It's stricter, but it's also more accurate.

**Reply by NormandoHall** [reply]
Thanks for the explanation @infovore

**Comment by NormandoHall** [comment]
Finally I wrapped with a `JSON.stringify()` the offending array, because in my case, I have a lot of ObjectIds, and not all are in the same fields.

I insist this is annoying and not dev friendly as SK claims
